test/Constraints/diagnostics.swift

```diff
@@ -1578,3 +1578,13 @@ func testAddMemberVsRemoveCall() {
   let b = Foo_74617()
   let c = (a + b).bar() // expected-error {{cannot call value of non-function type 'Float'}} {{22-24=}}
 }
+
+func testArrayLiteralMetatypeMismatch() {
+  // Make sure we can correctly turn 'T' into a hole here.
+  func foo<T>(_ xs: T.Type, _: Int) -> T {} // expected-note {{in call to function 'foo'}}
+  func foo(_ i: Int) {}
+  let x = foo([], i)
+  // expected-error@-1 {{generic parameter 'T' could not be inferred}}
+  // expected-error@-2 {{cannot convert value of type '[Any]' to expected argument type '_.Type'}}
+  // FIXME: The above should say `T.Type` not `_.Type` ^
```

Comment:
I think this is a bug in `FailureDiagnostic::resolveType`, might be easy to fix.

Reply:
Yeah the issue is that we need the hole's locator, but that currently requires iterating over the equivalence class, which we don't have in the Solution. So either we need to stick the locator on the type variable itself, or stick a mapping in the Solution

Reply:
Oh, I see.

Reply:
The actual type being diagnosed is part of the fix though, right? Unless I'm misunderstanding what you mean by transformation

Reply:
We never want to print type variables so maybe replacing them with an underlying generic parameter should be done as part of forming a solution instead of delaying it to diagnostics like we do right now.